src/dune_action/dune_action.ml

```diff
@@ -0,0 +1,130 @@
+module Protocol = Protocol
+module Path = Path
+open Protocol
+
+module Fs : sig
+  val read_directory : string -> (string list, Unix.error) Stdune.Result.t
+
+  val read_file : string -> (string, Unix.error) Stdune.Result.t
+
+  val write_file : string -> string -> (unit, Unix.error) Stdune.Result.t
+end = struct
+  let catch_unix_error_exceptions f =
+    try Stdune.Ok (f ()) with Unix.Unix_error (e, _, _) -> Stdune.Error e
+
+  let read_directory =
+    let rec loop dh acc =
+      match Unix.readdir dh with
+      | "."
+       |".." ->
+        loop dh acc
+      | s -> loop dh (s :: acc)
+      | exception End_of_file -> acc
+    in
+    fun path ->
+      catch_unix_error_exceptions (fun () ->
+        let dh = Unix.opendir path in
+        Stdune.Exn.protect
+          ~f:(fun () -> loop dh [])
+          ~finally:(fun () -> Unix.closedir dh))
+
+  let read_file path =
+    catch_unix_error_exceptions (fun () ->
+      Stdune.Io.String_path.read_file path)
+
+  let write_file path data =
+    catch_unix_error_exceptions (fun () ->
+      Stdune.Io.String_path.write_file path data)
+end
+
+(* TODO jstaron: Store targets to support dynamic target discovery! *)
+module Stage = struct
+  type 'a t =
+    { action : unit -> 'a
+    ; dependencies : Dependency.Set.t
+    }
+
+  let map (t : 'a t) ~f = { t with action = (fun () -> f (t.action ())) }
+
+  let both (t1 : 'a t) (t2 : 'b t) =
+    { action = (fun () -> (t1.action (), t2.action ()))
+    ; dependencies = Dependency.Set.union t1.dependencies t2.dependencies
+    }
+end
+
+(* TODO jstaron: Use ideas from
```

Comment:
To me it looks like your definition below is already like what they propose. I'm not sure what's left "TODO" here.